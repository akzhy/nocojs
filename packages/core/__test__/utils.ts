import path from 'path';
import { TransformOptions } from '..';
import { PreviewOptions } from '..';
import sharp from 'sharp';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';

export const defaultTransformOptions: TransformOptions = {
  cacheFileDir: path.join(import.meta.dirname, '.cache'),
  publicDir: path.join(import.meta.dirname, 'public'),
  logLevel: 'verbose',
};

interface GetInputProps {
  url?: string;
  previewOptions?: PreviewOptions;
}

export const getInput = (props?: GetInputProps | GetInputProps[]): string => {
  if (Array.isArray(props)) {
    const previewStatements = props
      .map((prop, i) => {
        const previewOptions = prop?.previewOptions ? `, ${JSON.stringify(prop.previewOptions)}` : '';
        const url = prop?.url || 'https://picsum.photos/id/237/200/300';

        return `const img${i} = preview("${url}"${previewOptions});`;
      })
      .join('\n');

    return `import { preview } from '@nocojs/client';

${previewStatements}`;
  }

  const previewOptions = props?.previewOptions ? `, ${JSON.stringify(props.previewOptions)}` : '';
  const url = props?.url || 'https://picsum.photos/id/237/200/300';

  return `import { preview } from '@nocojs/client';

let img = preview("${url}"${previewOptions});`;
};

export const base64ToSharpImage = (base64: string) => {
  const base64Data = base64.replace(/^data:image\/\w+;base64,/, '');
  const buffer = Buffer.from(base64Data, 'base64');

  return sharp(buffer);
};

export const numbersAreWithinPercent = (num1: number, num2: number, percent: number) => {
  const diff = Math.abs(num1 - num2);
  const max = Math.max(Math.abs(num1), Math.abs(num2));
  return diff <= (percent / 100) * max;
};

export const getDominantColor = async (image: sharp.Sharp) => {
  const { dominant } = await image.stats();
  return {
    rgb: dominant,
    string: `rgb(${dominant.r}, ${dominant.g}, ${dominant.b})`,
  };
};

export async function isImageSingleColor(sharpInstance: sharp.Sharp): Promise<boolean> {
  const { data, info } = await sharpInstance
    .clone()
    .resize(1, 1, { fit: 'fill' })
    .raw()
    .toBuffer({ resolveWithObject: true });

  const pixelValue = Array.from(data);

  const { data: originalData } = await sharpInstance.clone().raw().toBuffer({ resolveWithObject: true });

  for (let i = 0; i < originalData.length; i += info.channels) {
    for (let c = 0; c < info.channels; c++) {
      if (originalData[i + c] !== pixelValue[c]) {
        return false;
      }
    }
  }
  return true;
}

export async function isFullyTransparent(sharpInstance: sharp.Sharp) {
  const { data, info } = await sharpInstance.ensureAlpha().raw().toBuffer({ resolveWithObject: true });

  const channels = info.channels;
  const alphaIndex = channels - 1;

  for (let i = alphaIndex; i < data.length; i += channels) {
    if (data[i] !== 0) {
      return false;
    }
  }
  return true;
}

export function verifyPreviewCall(code: string) {
  const ast = parse(code, {
    sourceType: 'module', // or "script" if not ESM
    plugins: ['jsx', 'typescript'], // add if needed
  });

  let found = false;
  let imageUpdated = false;

  traverse(ast, {
    CallExpression(path) {
      const callee = path.node.callee;

      if (callee.type === 'Identifier' && callee.name === 'preview') {
        found = true;
        const args = path.node.arguments;
        const firstArg = args[0];

        if (firstArg.type === 'StringLiteral' && firstArg.value.startsWith('data:')) {
          imageUpdated = true;
        }

        path.stop(); // Stop early
      }
    },
  });

  return {
    found,
    imageUpdated,
  };
}
