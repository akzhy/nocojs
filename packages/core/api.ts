import * as path from 'path';
import {
  PlaceholderImageOutputKind,
  transform as rustTransform,
  getPlaceholder as rustGetPlaceholder,
  LogLevel,
  Log,
  GetPlaceholderOptions as RustGetPlaceholderOptions,
} from './index';

const placeholderTypeToEnum = {
  normal: PlaceholderImageOutputKind.Normal,
  blurred: PlaceholderImageOutputKind.Blurred,
  grayscale: PlaceholderImageOutputKind.Grayscale,
  'dominant-color': PlaceholderImageOutputKind.DominantColor,
  'average-color': PlaceholderImageOutputKind.AverageColor,
  transparent: PlaceholderImageOutputKind.Transparent,
} as const;

const logLevelTypeToEnum = {
  error: LogLevel.Error,
  info: LogLevel.Info,
  none: LogLevel.None,
  verbose: LogLevel.Verbose,
};

export type PlaceholderType = keyof typeof placeholderTypeToEnum;

export type LogLevelType = keyof typeof logLevelTypeToEnum;

export interface PreviewOptions {
  placeholderType?: PlaceholderType;
  replaceFunctionCall?: boolean;
  cache?: boolean;
  width?: number;
  height?: number;
  wrapWithSvg?: boolean;
}

export interface TransformOptions extends PreviewOptions {
  publicDir?: string;
  cacheFileDir?: string;
  logLevel?: LogLevelType;
  sourcemapFilePath?: string;
}

export const transform = async (
  code: string,
  filePath: string,
  options?: TransformOptions,
): Promise<{
  code: string;
  map: string | null;
  logs: Log[];
}> => {
  try {
    const result = await rustTransform(code, filePath, {
      placeholderType: options?.placeholderType
        ? placeholderTypeToEnum[options.placeholderType]
        : PlaceholderImageOutputKind.Normal,
      replaceFunctionCall: options?.replaceFunctionCall ?? true,
      cache: options?.cache ?? true,
      publicDir: options?.publicDir ?? path.join(process.cwd(), 'public'),
      cacheFileDir: options?.cacheFileDir ?? path.join(process.cwd(), '.nocojs'),
      logLevel: options?.logLevel ? logLevelTypeToEnum[options.logLevel] : LogLevel.Error,
      width: options?.width,
      height: options?.height,
      sourcemapFilePath: options?.sourcemapFilePath,
      wrapWithSvg: options?.wrapWithSvg ?? true,
    });

    if (!result) {
      return {
        code,
        map: null,
        logs: [],
      };
    }

    return {
      code: result.code,
      map: result?.sourcemap ?? null,
      logs: result.logs ?? [],
    };
  } catch (error) {
    console.error(`[nocojs] Error during transformation: ${error} file: ${filePath}`);
    return {
      code,
      map: null,
      logs: [],
    };
  }
};

export interface GetPlaceholderOptions extends Omit<RustGetPlaceholderOptions, 'placeholderType'> {
  placeholderType?: PlaceholderType;
}

export const getPlaceholder = async (url: string, options?: GetPlaceholderOptions) => {
  return rustGetPlaceholder(url, {
    ...options,
    placeholderType: options?.placeholderType
      ? placeholderTypeToEnum[options.placeholderType]
      : PlaceholderImageOutputKind.Normal,
  });
};
