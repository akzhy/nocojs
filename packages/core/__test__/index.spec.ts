import { readFile } from 'fs/promises';
import path from 'path';
import { describe, expect, test } from 'vitest';
import { transform } from '../api';
import { checkPreviewImage, defaultTransformOptions, getCacheFileDirName, getInput } from './utils';

describe('Basic Transform Tests', () => {
  test('transforms with no options', async () => {
    const input = getInput();
    const result = await transform(input, 'index.ts');
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('transforms with custom cache dir', async () => {
    const input = getInput();
    const result = await transform(input, 'index.ts', defaultTransformOptions);

    const cacheFilePath = path.join(defaultTransformOptions.cacheFileDir!, 'cache.db');

    const cacheFileExists = await readFile(cacheFilePath);
    expect(cacheFileExists).toBeDefined();

    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('transforms with custom publicDir', async () => {
    const input = getInput({
      url: '/good_boy_4x5.jpg',
    });
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      publicDir: path.join(__dirname, 'static'),
    });

    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('ignores invalid URLs and Paths', async () => {
    const input = getInput([
      {
        url: '/invalid-url.jpg',
      },
      {
        url: 'file:///invalid-path.jpg',
      },
      {
        url: 'https://example.com/invalid-image.jpg',
      },
      {
        url: '/good_boy_4x5.jpg',
      },
    ]);
    const result = await transform(input, 'index.ts', defaultTransformOptions);

    const matches = result.code.match(/const img\d*\s*=\s*(.*)/gm);
    if (!matches) {
      return;
    }

    expect(checkPreviewImage(matches[0])).toBe(false);
    expect(checkPreviewImage(matches[1])).toBe(false);
    expect(checkPreviewImage(matches[2])).toBe(false);
    expect(checkPreviewImage(matches[3])).toBe(true);
  }, 20000);
});

describe('Global placeholderType option tests with remote image', () => {
  test('placeholderType - normal', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'normal',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('placeholderType - normal', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'blurred',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('placeholderType - average-color', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'average-color',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('placeholderType - dominant-color', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'dominant-color',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('placeholderType - grayscale', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'grayscale',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });

  test('placeholderType - transparent', async () => {
    const cacheFileDir = getCacheFileDirName();
    const input = getInput();
    const result = await transform(input, 'index.ts', {
      ...defaultTransformOptions,
      placeholderType: 'transparent',
      cacheFileDir,
    });
    expect(checkPreviewImage(result.code)).toBeTruthy();
  });
});
