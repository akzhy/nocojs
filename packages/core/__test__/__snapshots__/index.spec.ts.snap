// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basic Transform Tests > ignores invalid URLs and Paths 1`] = `
"import { preview } from "@nocojs/client";
const img0 = preview("/invalid-url.jpg");
const img1 = preview("file:///invalid-path.jpg");
const img2 = preview("https://example.com/invalid-image.jpg");
const img3 = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAUCAIAAAALACogAAADp0lEQVR4Ae3AA6AkWZbG8f937o3IzKdyS2Oubdu2bdu2bdu2bWmMnpZKr54yMyLu+Xa3anqmhztr1a/ahuR+pnGZzHMyAFRI0wDZgACwDcnzQYVBYBsnNAAQkMnzsKhMKyQ84YQGYGOjxMFzklTJJTRlw41skHZT2l4LgVBBgkCJozLu0SbnpGzkSEut9pkf++ZP+OC7nvZ3n/f9P+ON6wCpAKhURtxC2WGxPnqfN36Dd3uta8/v7213pz7vx/6I2RlNiT20aRza0TTVb/39+/7qtvv+7Gm33jn51X7pM//yqav3fod3ZfWDX/0TT/ntP3nF8im/e9/+qrX2Kg950B17h7urtT70h/9iJh+u2mMffN3Jjf7vPunhf/bXi9d+1Qd/zy/8+c/+zi//4vCwn/yln32Nl3vVGKff+Is/u/aGB8VLX3vs0deeesSZ4zOVs5cuTcP4mV/8hU+59W+vnZdbf+yTnvh3f1aW+7/9O7/xkBuOv+VrveorPeza0r3Y6/z6b/7cbXfdtj/Vv/vz39r7s1993ON/55f/+ujlHzF/7Ve84RdWL3PfrU85uudpf/i3/zCM035blAe/5tucOHH93tHq0u6546cf9KaL373jzt2Hvuq7PPkv/+rR108Pfszr8qBXfsmXf40T1z14c7H96IfeVN/xdV5lYz47t3e06OJwuY5bH3J+ee7aZ/z0rHD3+eXLHR8f82KP2NyYn5rHia0up1GPv3P38XedXS+HF7vldK/2Ax/62rvn73rw9Yuze9NT7rj4Gi/2oLf6uj+//cLhU85duOb4DbP5ls5e3L39/PpSHluvxkHl8PDwZz7gQcevuamW6SVf+93Gv/7BxYf94eaikquHnd5x6eJJF7eOdHLMuOWaxYnNbtnWj7hx6+Pe7c3f6usf/5Jv+QF/c/vhw3bKpLK5ff3J01s/+5d3xt8/7Z7TW/zOn/7Vbz91efPxPNPNH/7hv/XXZ/Pgabc+6am7Fw79q1/9Ucfz4MlPfvLP/t7TcjXU2/b0c39+3xOeeudUF9985+aDtpVHew9908/45q/6qhPX3/Kh3/m3Wz0/+8u/uXXtQ2PzhNfr8opv+A7v/nqPfoUXe8TT7z56+1d6yN33nHvQLQ/668ff8ZZv+5bXnLrxx37yl+5cdWduuuWWWx6yHt3FVN7mgz7rsdcvau0edfP8wr47D8+49+wd913a3d275vT27Mwtd/7KV9z2d7/7sg/XU3/9Jx7rf/hHJDSs73ctbHMAAAAASUVORK5CYII=";
"
`;

exports[`Basic Transform Tests > transforms with custom cache dir 1`] = `
"import { preview } from "@nocojs/client";
let img = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAYCAIAAAB8wupbAAAEo0lEQVR4AQGYBGf7ALGfh9TCqsWzm7ejiLigholxV8y0msy0msqymLCYfqWNc62ZfpqGa7CYfraehKiQdgDYx628r43ArJPGsprQvKSmknfTvaW4o4S1ooTawqjmzrTKuKLCqo6lkYajkH+nj3MABgcLAAUAAgABDwoOBAADBQIAEBEWAAMGn456xKyS38etjYN3AwAHCQoFwqmLmoBvAAICAgUFBQQEBAUFBQAAAAEBAQUECgQEBAEABRQPCwIAAQABAAABAJ2EcKyUeLKZgwACAgIGBgYGBgYHBwcFAwQHAwQZEhoWFhYMDAwAAAAGBgYFBQU0MzgKCxACAwgLAgAABwcJDQwREA8UKysrAgIEMzI3AgEGFRQZDAwMCgoKAQEBBAQEEhEWERETDAwMDg4OAAYFCgUECRgXHA0MERIRFggHDAgDCUdARyEgJR4dIhUUGRUUGQICAgMDAxcXFwAAAAAREBUjIiciISYWFRpAP0RFREkbFhxJQklGRUpAP0QeHSI0MzgUFBQBAQEEBAQODg4ADAsQExIXKCcsJyYrIiEmAAAEBwYLGxofCAcMHh0iCwoPLSwxAQEBDAwMAwMDAAAAABUUGRMSFxcWGwIBBjAvNBkYHQAAAEpFSTErLQMDBRcWGysqLwUDBE5MTQEBAQAAAAAeHSIiISYGBQovKC8GCg0DAgcDBAY3LCo8Li0zNDYYFxw6OT4QEBAZGRkAAgEABAoAFhUaHx4jCgkOTEVMKCktAAAEDAsJAwQJYFRWMjM1OTc8FBMYBgYGBAQEBAABBgAAAAcGCxkYHQYFCg0IDggGCwUECQUJCBIHCyImKTw6Py8qMAwLEAkJCQICAg8KBEw6JgADAwUJCA0WFRodHCEWFhgEBAQAAAAVFRUJCQlOQ0keHSIFBAkWFRoJCQsmIheMemIAHx4jCQgNJiUqFBMYEhEWDQwRDw4UAgICBQECTkpLKSUmGhkfERERISIkBAMBDwMHAAcBDRkOEi4tNRMTFRAPFBoZHh8eIxUUGQ8LDFpWVwwICRoaGgAAAAICAgkTFDIlHACskX57alhFOy86OEUhICUODRIUExgLCg8FAQJNSUowLC0gICAZGRkBAQFKPDFsVzoAnYt1vayamoZuHBcbKikuNDM4Hx4kAQEBAwMDAAAACAgIHBwcAAAFKRwUkoJrspmFAMKuk7yojb6qj6SShj0+QgwLEA8OFA8LDAYCABMSEAEBARAPFAcGBFU7Krujia6WfAC4oIbAqI7OvKbq1L0HBQgVFBkLChAAAAAIBQxLODE3LywLBgwqIx2vlXysk32bg2kAqI95qZR5tJ+Kr5uDrZaEKSkzBAMIAAAEAgIEZ1VBnYlupJB4n4dtpI90qJB4jHRaANK5o62Zfr+qlb6qkpB6YgEECQgHDA0MEQAABI99Z825nqyXfK6WfMevlbefhaaOdAChiHK+pY/OtZ+qkXvGtZsLBwAXFhsCAQYFAwSpmYC4qI+/qZLHrpiwl4GtnYS3nogAk35hrJd4pI9wnYhpv6qNnYh1mId9CAAAk4N0tqGCu6aHoItsoY5ukn1er59+mYRl5nIONInhO2cAAAAASUVORK5CYII=";
"
`;

exports[`Basic Transform Tests > transforms with custom publicDir 1`] = `
"import { preview } from "@nocojs/client";
let img = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAUCAIAAAALACogAAADp0lEQVR4Ae3AA6AkWZbG8f937o3IzKdyS2Oubdu2bdu2bdu2bWmMnpZKr54yMyLu+Xa3anqmhztr1a/ahuR+pnGZzHMyAFRI0wDZgACwDcnzQYVBYBsnNAAQkMnzsKhMKyQ84YQGYGOjxMFzklTJJTRlw41skHZT2l4LgVBBgkCJozLu0SbnpGzkSEut9pkf++ZP+OC7nvZ3n/f9P+ON6wCpAKhURtxC2WGxPnqfN36Dd3uta8/v7213pz7vx/6I2RlNiT20aRza0TTVb/39+/7qtvv+7Gm33jn51X7pM//yqav3fod3ZfWDX/0TT/ntP3nF8im/e9/+qrX2Kg950B17h7urtT70h/9iJh+u2mMffN3Jjf7vPunhf/bXi9d+1Qd/zy/8+c/+zi//4vCwn/yln32Nl3vVGKff+Is/u/aGB8VLX3vs0deeesSZ4zOVs5cuTcP4mV/8hU+59W+vnZdbf+yTnvh3f1aW+7/9O7/xkBuOv+VrveorPeza0r3Y6/z6b/7cbXfdtj/Vv/vz39r7s1993ON/55f/+ujlHzF/7Ve84RdWL3PfrU85uudpf/i3/zCM035blAe/5tucOHH93tHq0u6546cf9KaL373jzt2Hvuq7PPkv/+rR108Pfszr8qBXfsmXf40T1z14c7H96IfeVN/xdV5lYz47t3e06OJwuY5bH3J+ee7aZ/z0rHD3+eXLHR8f82KP2NyYn5rHia0up1GPv3P38XedXS+HF7vldK/2Ax/62rvn73rw9Yuze9NT7rj4Gi/2oLf6uj+//cLhU85duOb4DbP5ls5e3L39/PpSHluvxkHl8PDwZz7gQcevuamW6SVf+93Gv/7BxYf94eaikquHnd5x6eJJF7eOdHLMuOWaxYnNbtnWj7hx6+Pe7c3f6usf/5Jv+QF/c/vhw3bKpLK5ff3J01s/+5d3xt8/7Z7TW/zOn/7Vbz91efPxPNPNH/7hv/XXZ/Pgabc+6am7Fw79q1/9Ucfz4MlPfvLP/t7TcjXU2/b0c39+3xOeeudUF9985+aDtpVHew9908/45q/6qhPX3/Kh3/m3Wz0/+8u/uXXtQ2PzhNfr8opv+A7v/nqPfoUXe8TT7z56+1d6yN33nHvQLQ/668ff8ZZv+5bXnLrxx37yl+5cdWduuuWWWx6yHt3FVN7mgz7rsdcvau0edfP8wr47D8+49+wd913a3d275vT27Mwtd/7KV9z2d7/7sg/XU3/9Jx7rf/hHJDSs73ctbHMAAAAASUVORK5CYII=";
"
`;

exports[`Basic Transform Tests > transforms with no options 1`] = `
"import { preview } from "@nocojs/client";
let img = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAYCAIAAAB8wupbAAAEo0lEQVR4AQGYBGf7ALGfh9TCqsWzm7ejiLigholxV8y0msy0msqymLCYfqWNc62ZfpqGa7CYfraehKiQdgDYx628r43ArJPGsprQvKSmknfTvaW4o4S1ooTawqjmzrTKuKLCqo6lkYajkH+nj3MABgcLAAUAAgABDwoOBAADBQIAEBEWAAMGn456xKyS38etjYN3AwAHCQoFwqmLmoBvAAICAgUFBQQEBAUFBQAAAAEBAQUECgQEBAEABRQPCwIAAQABAAABAJ2EcKyUeLKZgwACAgIGBgYGBgYHBwcFAwQHAwQZEhoWFhYMDAwAAAAGBgYFBQU0MzgKCxACAwgLAgAABwcJDQwREA8UKysrAgIEMzI3AgEGFRQZDAwMCgoKAQEBBAQEEhEWERETDAwMDg4OAAYFCgUECRgXHA0MERIRFggHDAgDCUdARyEgJR4dIhUUGRUUGQICAgMDAxcXFwAAAAAREBUjIiciISYWFRpAP0RFREkbFhxJQklGRUpAP0QeHSI0MzgUFBQBAQEEBAQODg4ADAsQExIXKCcsJyYrIiEmAAAEBwYLGxofCAcMHh0iCwoPLSwxAQEBDAwMAwMDAAAAABUUGRMSFxcWGwIBBjAvNBkYHQAAAEpFSTErLQMDBRcWGysqLwUDBE5MTQEBAQAAAAAeHSIiISYGBQovKC8GCg0DAgcDBAY3LCo8Li0zNDYYFxw6OT4QEBAZGRkAAgEABAoAFhUaHx4jCgkOTEVMKCktAAAEDAsJAwQJYFRWMjM1OTc8FBMYBgYGBAQEBAABBgAAAAcGCxkYHQYFCg0IDggGCwUECQUJCBIHCyImKTw6Py8qMAwLEAkJCQICAg8KBEw6JgADAwUJCA0WFRodHCEWFhgEBAQAAAAVFRUJCQlOQ0keHSIFBAkWFRoJCQsmIheMemIAHx4jCQgNJiUqFBMYEhEWDQwRDw4UAgICBQECTkpLKSUmGhkfERERISIkBAMBDwMHAAcBDRkOEi4tNRMTFRAPFBoZHh8eIxUUGQ8LDFpWVwwICRoaGgAAAAICAgkTFDIlHACskX57alhFOy86OEUhICUODRIUExgLCg8FAQJNSUowLC0gICAZGRkBAQFKPDFsVzoAnYt1vayamoZuHBcbKikuNDM4Hx4kAQEBAwMDAAAACAgIHBwcAAAFKRwUkoJrspmFAMKuk7yojb6qj6SShj0+QgwLEA8OFA8LDAYCABMSEAEBARAPFAcGBFU7Krujia6WfAC4oIbAqI7OvKbq1L0HBQgVFBkLChAAAAAIBQxLODE3LywLBgwqIx2vlXysk32bg2kAqI95qZR5tJ+Kr5uDrZaEKSkzBAMIAAAEAgIEZ1VBnYlupJB4n4dtpI90qJB4jHRaANK5o62Zfr+qlb6qkpB6YgEECQgHDA0MEQAABI99Z825nqyXfK6WfMevlbefhaaOdAChiHK+pY/OtZ+qkXvGtZsLBwAXFhsCAQYFAwSpmYC4qI+/qZLHrpiwl4GtnYS3nogAk35hrJd4pI9wnYhpv6qNnYh1mId9CAAAk4N0tqGCu6aHoItsoY5ukn1er59+mYRl5nIONInhO2cAAAAASUVORK5CYII=";
"
`;
